### DISCUSSION TITLE
rotate in 2.x not working (+ transitions and on_leave not working)

### DISCUSSION THREAD

**Discussion by FlorianReiterer** [savonet/liquidsoap#2081] [User Support]
Rotate does not create correct weights in 2.x liquidsoap's
This is supposed to be fixed in 2.01, see #1484
@smimram:

> I can confirm that this is broken in 1.4.3
> ```
> songs   = sine(duration=1., 440.)
> jingles = sine(duration=1., 880.)
> radio   = rotate(weights=[1, 3], [jingles, songs])
> output.alsa(fallible=true, radio)
> ```
> and that this is working in 2.0
> ```
> songs   = single("synth:frequency=440,d=1")
> jingles = single("synth:frequency=880,d=1")
> radio   = rotate(weights=[1, 3], [jingles, songs])
> output.alsa(fallible=true, radio)
> ```

but rotate for a playlist does not work for me. My complete script here:
```
settings.frame.audio.channels.set(2)
settings.frame.audio.samplerate.set(96000)

l = playlog() 
def check(r) 
m = request.metadata(r) 
if l.last(m) < 6900. then 
log.info("Rejecting #{m['filename']} (played #{l.last(m)}s ago).")
 false 
else
 l.add(m)
 true
 end 
end

#Live Input
live = input.harbor("live", port=18888, password="hackme")

Musik = playlist(reload_mode="watch",check_next=check,"/media/pi/Elements 4TB/FTP Mother Earth/Musik")
Musik = crossfade(smart=true, width=4., high=-25., medium=-39., Musik)
Klassik = playlist(reload_mode="watch",check_next=check,"/media/pi/Elements 4TB/FTP Mother Earth/Klassik")
show = ("/media/pi/Elements 4TB/FTP Mother Earth/Sendungen/show.flac")
music = rotate (weights = [12, 1],[Musik, Klassik])
music.on_leave(music.skip)

showtime = request.queue()
thread.when({3w18h or 7w11h}, {showtime.push.uri(show)})

def to_show(a, b) 
  add(normalize=false, [
    fade.out(a),
    sequence([blank(duration=2.), (b:source)]),
  ])
end

radio = fallback(track_sensitive=false, transitions=[to_show], [showtime, music])
radio = fallback(track_sensitive=false, transitions=[to_show], [live, radio])
output.icecast(%ogg(%flac(samplerate=96000, channels=2, compression=5, bits_per_sample=24)), fallible = true,icy_metadata = "false", mount="source", password="hackme", port=18800, buffer(buffer=5.,(radio)))
output.icecast(%mp3(bitrate=256,samplerate=44100), format="audio/mpeg", fallible=true, icy_metadata = "true", mount="source.mp3", password="hackme", port=18800, clock(buffer(buffer=5.,(radio))))
```
other issues with this script are that the transition is not working (not fading out source a = the playlist that does not rotate properly)
and that `music.on_leave `does not skip.
Maybe all this is related, therefore the whole script and the multiple issues in one post.
The rotate is vital for my station. It is already on air because of all the awesome features in 2.x, but my station is not playing any classical music anymore, which is ... not good.
Thank you for looking into it.

regards, Florian
tested versions:
`Liquidsoap 2.1.0+git@6a6e0c9e`
```
version       2.0.0
repository    default
url.src:      "https://github.com/savonet/liquidsoap/releases/download/v2.0.0/liquidsoap-2.0.0.tar.bz2"
```
```
version      2.0.0
repository   default
pin          file:///home/pi/Downloads/liquidsoap-2.0.1

```
related topics: #1484, #1651, #1990, #1930

**Comment by FlorianReiterer** [comment]
THIS `rotate` works:
```
Musik = playlist("/media/pi/Elements 4TB/rotate/Primus")

Klassik = playlist("/media/pi/Elements 4TB/rotate/Bach")
radio = rotate (weights = [2, 1],[Musik, Klassik])

output.icecast(%mp3(bitrate=256,samplerate=44100), format="audio/mpeg", 
  fallible=true, mount="source.mp3", password="hackme", port=18800, 
  buffer(buffer=5.,(radio)))
```
and this `crossfade` breaks it:
```
Musik = playlist("/media/pi/Elements 4TB/rotate/Primus")
Musik = crossfade(smart=true, width=4., high=-25., medium=-39., Musik)
Klassik = playlist("/media/pi/Elements 4TB/rotate/Bach")
radio = rotate (weights = [2, 1],[Musik, Klassik])

output.icecast(%mp3(bitrate=256,samplerate=44100), format="audio/mpeg", 
  fallible=true, mount="source.mp3", password="hackme", port=18800, 
  buffer(buffer=5.,(radio)))
```

**Comment by toots** [comment]
Hi!

I believe that this is to be expected. This line:
```
Musik = crossfade(smart=true, width=4., high=-25., medium=-39., Musik)
```
Will blur the track marks between songs in this playlist. Therefore, the `rotate` operator does not know when to switch away from the source, because it is done on track marks.

The solution is to add the crossfade _after_ the rotate.

As for `on_leave`, this is an internal function and shouldn't be used. What exactly are you trying to do with it?

Lastly, transitions won't work properly with a live source because, when switching away, there is no more data to use to compute the transition. There might be ways to mitigate that using `buffer` but I would recommend to try it before putting it to production.

I am moving this to the discussions, hopefully this response helps!

**Comment by FlorianReiterer** [comment]
Thanks, that explains rotate.

with `on_leave` i wanted skip the song from the playlist that has been left when a switch occured, so that a new track gets played when live is over. I'll read that up again.

So, a transition to a live show will always interrupt the playing song (tracks_sensitive=false) with a hard cut?

**Reply by toots** [reply]
Yeah, you should do the skip in the transition.

No, a transition away from a playlist should be able to buffer and fade out the current song. If that's not working then this would be a bug!

**Comment by FlorianReiterer** [comment]
well, thats what i m seeing - transition away from playlist to live with this:
```
def to_show(a, b) 
  add(normalize=false, [
    fade.out(a),
    sequence([blank(duration=2.), (b:source)]),
  ])
end
```
